packages/scrapers-v2/src/runtime/logger.ts

```diff
@@ -0,0 +1,175 @@
+      stats.recordField("status", "filtered");
+      stats.recordField("filtered", result.err[0].actual);
+      if (result.err[0].actual === CatalogEntryType.Unknown) {
+        console.log("entry with unknown type:", id.toString());
+      }
+      continue;
+    }
+
+    stats.recordField("status", "error");
+    stats.recordField("stage failures", trace[trace.length - 1]);
+
+    for (const err of result.err) {
+      if (err instanceof Error) {
+        stats.recordError(err, id);
+      } else {
+        stats.recordError(new Error(`non-error value: ${err}`), id);
+      }
+    }
+  }
+
+  stats.print();
+};
+
+/**
+ * Allows for "recording" fields (and errors) to print a breakdown of the
+ * different values at the end.
+ *
+ * Each field value will be added to a tally. The # of occurrences of each value
+ * for a field are then displayed when `print()` is called.
+ *
+ * Errors are also tallied. Error grouping is done by comparing stacktrace. This
+ * doesn't quite work for async stacktraces, so sometimes two of the same error
+ * are displayed separately.
+ */
+class StatsLogger {
+  // field -> value -> count
+  private fields: Record<string, Map<any, number>> = {};
+  // message -> list -> stacktrace
+  private errors: Map<
+    string,
+    Array<{ err: Error; count: number; annot: string; entryIds: URL[] }>
+  > = new Map();
+
+  /**
+   * Records a field and its value, with the goal of printing the counts for
+   * each different value the field has at the end.
+   *
+   * @param field The name of the field
+   * @param value The value of the field
+   */
+  recordField(field: string, value: any) {
+    if (field === "errors") {
+      throw new Error(
+        "cannot use 'errors' as a field key, use a different name"
+      );
+    }
+    this.record(field, value);
+  }
+
+  /**
+   * Records an error for a specific entry. Error uniqueness is determined by
+   * stack trace. This doesn't quite work for async stacktraces, so sometimes
+   * two of the same error are displayed separately.
+   *
+   * @param err     The error
+   * @param entryId The entry URL
+   */
+  recordError(err: Error, entryId: URL) {
+    const key = err.message ?? "had no stacktrace";
+    const storedErrors = this.errors.get(key) ?? [];
+    for (const stored of storedErrors) {
+      // if the stacktrace matches, increment the count
+      if (err.stack === stored.err.stack) {
+        stored.count += 1;
+        stored.entryIds.push(entryId);
+        this.record("errors", stored.annot);
+        return;
+      }
+    }
+    const id = storedErrors.length === 0 ? "" : ` #${storedErrors.length}`;
+    const annot = `${err.message}${id}`;
+    // stacktrace didn't match, so add a new stacktrace entry for this error message
+    storedErrors.push({ err, count: 1, annot, entryIds: [entryId] });
+    this.errors.set(key, storedErrors);
+    this.record("errors", annot);
+  }
+
+  private record(field: string, value: any) {
```

Comment:
smol comment explaining that this is the func that increments count for a field?

Reply:
Yep